File: frontend/src/api/campaigns.ts
import { api } from './client';

export interface Campaign {
  id: number;
  name: string;
  city: string;
  category: string;
  description: string | null;
  price_per_task: number;
  status: 'draft' | 'active' | 'completed';
  created_at: string;
  published_at: string | null;
  total_tasks: number;
  completed_tasks: number;
  pending_tasks: number;
  in_progress_tasks: number;
  rejected_tasks: number;
}

export interface TaskReport {
  id: number;
  task_id: number;
  address: string;
  status: 'available' | 'in_progress' | 'pending_review' | 'completed' | 'rejected';
  photo_url: string | null;
  photo_taken_at: string | null;
  check_result: Record<string, unknown> | null;
  client_confirmed: boolean | null;
  rejection_reason: string | null;
  executor_phone: string | null;
  created_at: string;
}

export const campaignsApi = {
  list: () => api.get<Campaign[]>('/campaigns').then((r) => r.data),
  get: (id: number) => api.get<Campaign>(`/campaigns/${id}`).then((r) => r.data),
  create: (data: {
    name: string;
    city: string;
    category: string;
    description?: string;
    price_per_task: number;
    addresses: string[];
  }) => api.post<Campaign>('/campaigns', data).then((r) => r.data),
  publish: (id: number) => api.post(`/campaigns/${id}/publish`).then((r) => r.data),
  reports: (id: number) =>
    api.get<TaskReport[]>(`/tasks/campaign/${id}/reports`).then((r) => r.data),
  moderate: (taskId: number, approved: boolean, rejectionReason?: string) => {
    const form = new FormData();
    form.append('approved', String(approved));
    if (rejectionReason) form.append('rejection_reason', rejectionReason);
    return api.post(`/tasks/${taskId}/moderate`, form).then((r) => r.data);
  },
};

export const authApi = {
  login: (email: string, password: string) =>
    api.post<{ access_token: string; user_id: number; name: string }>(
      '/auth/login',
      { email, password }
    ).then((r) => r.data),
  register: (data: {
    email: string;
    password: string;
    name: string;
    company_name?: string;
    phone?: string;
  }) =>
    api.post<{ access_token: string; user_id: number; name: string }>(
      '/auth/register',
      data
    ).then((r) => r.data),
};

export const notificationsApi = {
  list: () =>
    api.get<{ id: number; title: string; body: string; is_read: boolean; created_at: string }[]>(
      '/notifications'
    ).then((r) => r.data),
  readAll: () => api.post('/notifications/read-all').then((r) => r.data),
};

File: backend/app/api/campaigns.py
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.api.deps import get_current_client
from app.core.database import get_db
from app.models.campaign import Campaign, CampaignStatus
from app.models.task import Task, TaskStatus
from app.models.user import Client
from app.services.geo import geocode_address

router = APIRouter(prefix="/campaigns", tags=["campaigns"])


class CampaignCreate(BaseModel):
    name: str
    city: str  # "Москва" — используется при геокодировании
    category: str
    description: str | None = None
    price_per_task: float = 150
    addresses: list[str]


class CampaignOut(BaseModel):
    id: int
    name: str
    city: str
    category: str
    description: str | None
    price_per_task: float
    status: str
    created_at: datetime
    published_at: datetime | None
    total_tasks: int = 0
    completed_tasks: int = 0
    pending_tasks: int = 0
    pending_review_tasks: int = 0
    rejected_tasks: int = 0
    in_progress_tasks: int = 0

    model_config = {"from_attributes": True}


def _campaign_out(c: Campaign, tasks: list) -> CampaignOut:
    return CampaignOut(
        id=c.id,
        name=c.name,
        city=c.city,
        category=c.category,
        description=c.description,
        price_per_task=float(c.price_per_task),
        status=c.status.value,
        created_at=c.created_at,
        published_at=c.published_at,
        total_tasks=len(tasks),
        completed_tasks=sum(1 for t in tasks if t.status == TaskStatus.completed),
        pending_tasks=sum(1 for t in tasks if t.status == TaskStatus.available),
        pending_review_tasks=sum(1 for t in tasks if t.status == TaskStatus.pending_review),
        in_progress_tasks=sum(1 for t in tasks if t.status == TaskStatus.in_progress),
        rejected_tasks=sum(1 for t in tasks if t.status == TaskStatus.rejected),
    )


@router.get("", response_model=list[CampaignOut])
async def list_campaigns(
    client: Client = Depends(get_current_client),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(Campaign)
        .where(Campaign.client_id == client.id)
        .options(selectinload(Campaign.tasks))
        .order_by(Campaign.created_at.desc())
    )
    return [_campaign_out(c, c.tasks) for c in result.scalars().all()]


@router.post("", response_model=CampaignOut, status_code=status.HTTP_201_CREATED)
async def create_campaign(
    data: CampaignCreate,
    client: Client = Depends(get_current_client),
    db: AsyncSession = Depends(get_db),
):
    campaign = Campaign(
        client_id=client.id,
        name=data.name,
        city=data.city.strip(),
        category=data.category,
        description=data.description,
        price_per_task=data.price_per_task,
        status=CampaignStatus.draft,
    )
    db.add(campaign)
    await db.flush()

    tasks = []
    for address in data.addresses:
        address = address.strip()
        if not address:
            continue
        coords = await geocode_address(address, city=data.city.strip())
        task = Task(
            campaign_id=campaign.id,
            address=address,
            lat=coords["lat"] if coords else None,
            lng=coords["lng"] if coords else None,
            status=TaskStatus.available,
        )
        db.add(task)
        tasks.append(task)

    await db.commit()
    await db.refresh(campaign)
    return _campaign_out(campaign, tasks)


@router.get("/{campaign_id}", response_model=CampaignOut)
async def get_campaign(
    campaign_id: int,
    client: Client = Depends(get_current_client),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(Campaign)
        .where(Campaign.id == campaign_id, Campaign.client_id == client.id)
        .options(selectinload(Campaign.tasks))
    )
    campaign = result.scalar_one_or_none()
    if not campaign:
        raise HTTPException(status_code=404, detail="Кампания не найдена")
    return _campaign_out(campaign, campaign.tasks)


@router.post("/{campaign_id}/publish")
async def publish_campaign(
    campaign_id: int,
    client: Client = Depends(get_current_client),
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(Campaign).where(Campaign.id == campaign_id, Campaign.client_id == client.id)
    )
    campaign = result.scalar_one_or_none()
    if not campaign:
        raise HTTPException(status_code=404, detail="Кампания не найдена")
    if campaign.status != CampaignStatus.draft:
        raise HTTPException(status_code=400, detail="Кампания уже опубликована")

    campaign.status = CampaignStatus.active
    campaign.published_at = datetime.now(timezone.utc)
    await db.commit()
    return {"ok": True}
